### PR TITLE
expr: Support json_merge_patch pushdown

### LINKED ISSUE #16770: support json_merge_patch pushdown to tikv
### Is your feature request related to a problem? Please describe:

https://dev.mysql.com/doc/refman/8.3/en/json-modification-functions.html#function_json-merge-patch

### PR DESCRIPTION
**Pull request by dbsid**
### What is changed and how it works?

Issue Number: Close #16770

What's Changed:

```commit-message

```

### Related changes

- [x] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
support json_merge_patch pushdown to tikv
```

### CODE REVIEW

File: components/tidb_query_datatype/src/codec/mysql/json/json_merge.rs [hunk 1]
```diff
@@ -103,6 +160,51 @@ fn merge_binary_object(objects: &mut Vec<JsonRef<'_>>) -> Result<Json> {
...
+// See `mergePatchBinaryJSON()` in TiDB `pkg/types/json_binary_functions.go`
+fn merge_patch_binary_object(target: JsonRef<'_>, patch: JsonRef<'_>) -> Result<Json> {
+    // translate the function from go to rust
+    if patch.get_type() != JsonType::Object {
+        return Ok(patch.to_owned());
+    }
+
+    if target.get_type() != JsonType::Object {
+        return Ok(patch.to_owned());
+    }
```
Comment (breezewish) [inline components/tidb_query_datatype/src/codec/mysql/json/json_merge.rs:172]: Not sure, looks like TiDB's code doesn't looks like in this way?

**Review by breezewish — COMMENTED** [review]
The rest LGTM

**Review by breezewish — APPROVED** [review]

**Review by gengliqi — APPROVED** [review]

**Pull request merged**